core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
+   * Cleanup Phase: In the cases where this reassignment has to override an ongoing reassignment.
+   *   . The ongoing reassignment is in phase A
+   *   . ORS denotes the original replica set, prior to the ongoing reassignment
+   *   . URS denotes the unnecessary replicas, ones which are currently part of the ongoing reassignment but will not be part of the new one
+   *   . OVRS denotes the overlapping replica set - replicas which are part of the ongoing reassignment and will be part of the overriding reassignment
+   *       (it is essentially (RS - ORS) - URS)
+   *   1 Set RS = ORS + OVRS, AR = OVRS, RS = [] in memory
+   *   2 Send LeaderAndIsr request with RS = ORS + OVRS, AR = [], RS = [] to all brokers in ORS + OVRS
+   *     (because the ongoing reassignment is in phase A, we know we wouldn't have a leader in URS)
+   *   3 Replicas in URS -> Offline (force those replicas out of ISR)
+   *   4 Replicas in URS -> NonExistentReplica (force those replicas to be deleted)
+   *   5 Update ZK with RS = ORS + OVRS, AR = OVRS, RS = []
    *
-   * For example, if OAR = {1, 2, 3} and RAR = {4,5,6}, the values in the assigned replica (AR) and leader/isr path in ZK
-   * may go through the following transition.
-   * AR                 leader/isr
-   * {1,2,3}            1/{1,2,3}           (initial state)
-   * {1,2,3,4,5,6}      1/{1,2,3}           (step 2)
-   * {1,2,3,4,5,6}      1/{1,2,3,4,5,6}     (step 4)
-   * {1,2,3,4,5,6}      4/{1,2,3,4,5,6}     (step 7)
-   * {1,2,3,4,5,6}      4/{4,5,6}           (step 8)
-   * {4,5,6}            4/{4,5,6}           (step 10)
+   * Phase A: Initial trigger (when RAR != ISR)
+   *   0. Update memory with RS = ORS + TRS, AR = TRS - ORS and RR = ORS - TRS
+   *   A1. Update ZK with RS = ORS + TRS,
+   *                      AR = TRS - ORS and
+   *                      RR = ORS - TRS.
+   *   A2. Send LeaderAndIsr request to every replica in ORS + TRS (with the new RS, AR and RR).
+   *       We do this by forcing an update of the leader epoch in zookeeper.
+   *   A3. Start new replicas AR by moving replicas in AR to NewReplica state.
    *
-   * Note that we have to update AR in ZK with RAR last since it's the only place where we store OAR persistently.
+   * Phase B: All of RAR have caught up with the leaders and are in ISR
+   *   0. Update memory with RS = ORS + TRS, AR = TRS - ORS and RR = ORS - TRS (needed in cases where RAR == ISR from the initial trigger)
+   *   B1. Move all replicas in TRS to OnlineReplica state.
+   *   B2. Set RS = TRS, AR = [], RR = [] in memory.
+   *   B3. If the leader is not in TRS, elect a new leader from TRS. If new leader needs to be elected from TRS, a LeaderAndIsr
+   *       will be sent. If not, then leader epoch will be incremented in zookeeper and a LeaderAndIsr request will be sent.
+   *       In any case, the LeaderAndIsr request will have RS = TRS. This will prevent the leader from adding any replica in
```

Comment:
The leader epoch needs to be incrementing in both cases, right?  If a new leader needs to be elected, and also if a new leader doesn't need to be elected.  So we should just add this as a step.